java/arcs/core/storage/referencemode/MessageQueue.kt

```diff
@@ -45,27 +47,47 @@ class MessageQueue(
     @ExperimentalCoroutinesApi
     suspend fun enqueue(message: Message): Boolean {
         require(message !is Message.Enqueued) { "Cannot enqueue an already-enqueued message." }
-        size.incrementAndGet().let {
-            if (!sizeChannel.isClosedForSend) sizeChannel.send(it)
-        }
         val deferred = CompletableDeferred<Boolean>(coroutineContext[Job.Key])
-        mutex.withLock { queue += message.toEnqueued(deferred) }
+        mutex.withLock { queue += QueueEntry.MessageEntry(message.toEnqueued(deferred)) }
         CoroutineScope(coroutineContext).launch { drainQueue() }
-        return deferred.await().also {
-            size.decrementAndGet().let {
-                if (!sizeChannel.isClosedForSend) sizeChannel.send(it)
-            }
-        }
+        return deferred.await()
     }
 
     private suspend fun drainQueue() {
         val messagesToProcess = mutex.withLock {
-            mutableListOf<Message>().also {
+            mutableListOf<QueueEntry>().also {
                 it.addAll(queue)
                 queue.clear()
             }
         }
-        messagesToProcess.forEach { processMessage(it); Unit }
+
+        messagesToProcess.forEach {
+            when (it) {
+                is QueueEntry.MessageEntry -> processMessage(it.message)
+                is QueueEntry.Flush -> Unit
+            }
+        }
+
+        mutex.withLock {
+            if (queue.isEmpty()) {
```

Comment:
Two or more `drainQueue` can be taking place concurrently for different message batches.
A bug here where the 1st batch gets done (so `queue` is empty) but the 2nd batch is still running `messagesToProcess.forEach`.

Reply:
Ahh, good point.

I noticed that two or more drainQueues might happen concurrently, and wonder if that's actually behavior we want?

But I hadn't thought about that in relation to this particular block, thanks for pointing it out.